ocs/ocp.py

```diff
@@ -63,14 +63,17 @@ def exec_oc_cmd(self, command):
 
         oc_cmd += command
         out = run_cmd(cmd=oc_cmd)
+
+        # TODO: this part causes yaml.safe_load() to fail
+
         # Removing any part of output which is out of the yaml/json
-        try:
-            if '{' in out:
-                if not out.startswith('{'):
-                    if ':' not in out[:out.index('{')]:
-                        out = out[out.index('{'):]
-        except ValueError:
-            pass
+        # try:
```

Comment:
Which command is failing? What is the output before this line?

Reply:
17:40:23 - MainThread - utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ebenahar/ebenahar-ocs/auth/kubeconfig rsh rook-ceph-tools-76c7d559b6-mfhm2 ceph osd lspools --format json-pretty




    @pytest.fixture()
    def create_ceph_block_pool(request):
        """
        Create a Ceph block pool
        """
        class_instance = request.node.cls
    
        def finalizer():
            """
            Delete the Ceph block pool
            """
            if class_instance.cbp_obj.get():
                class_instance.cbp_obj.delete()
    
        request.addfinalizer(finalizer)
    
        cbp_data = defaults.CEPHBLOCKPOOL_DICT.copy()
        cbp_data['metadata']['name'] = helpers.create_unique_resource_name(
            'test', 'cbp'
        )
        class_instance.cbp_obj = helpers.create_resource(**cbp_data, wait=False)
        assert class_instance.cbp_obj, (
            f"Failed to create block pool {cbp_data['metadata']['name']}"
        )
>       assert helpers.verify_block_pool_exist(class_instance.cbp_obj.name), (
            f"Block pool {class_instance.cbp_obj.name} does not exist"
        )

tests/fixtures.py:59: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/helpers.py:101: in verify_block_pool_exist
    pools = ct_pod.exec_ceph_cmd('osd lspools')
resources/pod.py:125: in exec_ceph_cmd
    out = self.exec_cmd_on_pod(ceph_cmd)
resources/pod.py:90: in exec_cmd_on_pod
    return self.ocp.exec_oc_cmd(rsh_cmd)
ocs/ocp.py:78: in exec_oc_cmd
    return yaml.safe_load(out)
ocs-venv/lib64/python3.7/site-packages/yaml/__init__.py:162: in safe_load
    return load(stream, SafeLoader)
ocs-venv/lib64/python3.7/site-packages/yaml/__init__.py:114: in load
    return loader.get_single_data()
ocs-venv/lib64/python3.7/site-packages/yaml/constructor.py:41: in get_single_data
    node = self.get_single_node()
ocs-venv/lib64/python3.7/site-packages/yaml/composer.py:39: in get_single_node
    if not self.check_event(StreamEndEvent):
ocs-venv/lib64/python3.7/site-packages/yaml/parser.py:98: in check_event
    self.current_event = self.state()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <yaml.loader.SafeLoader object at 0x7fcd6b4d6f60>

    def parse_document_start(self):
    
        # Parse any extra document end indicators.
        while self.check_token(DocumentEndToken):
            self.get_token()
    
        # Parse an explicit document.
        if not self.check_token(StreamEndToken):
            token = self.peek_token()
            start_mark = token.start_mark
            version, tags = self.process_directives()
            if not self.check_token(DocumentStartToken):
                raise ParserError(None, None,
                        "expected '<document start>', but found %r"
                        % self.peek_token().id,
>                       self.peek_token().start_mark)
E               yaml.parser.ParserError: expected '<document start>', but found ','
E                 in "<unicode string>", line 4, column 6:
E                       },
E                        ^

ocs-venv/lib64/python3.7/site-packages/yaml/parser.py:174: ParserError

Reply:
ok that was a long debugging:
So this code that you commented came from the output of the command:
oc -n openshift-storage --kubeconfig /home/ratamir/ocs_automation/auth/kubeconfig rsh rook-ceph-tools-75867b94fc-q48cp rados --no-log-to-stderr -b 4194304 -p test_pool506 bench 10 write --no-cleanup  --format json-pretty
which is from test_bench.py 
So it seems like this command returning output that looks like:
```
# oc -n openshift-storage --kubeconfig /home/ratamir/ocs_automation/auth/kubeconfig rsh rook-ceph-tools-75867b94fc-q48cp rados --no-log-to-stderr -b 4194304 -p test_pool506 bench 10 write --no-cleanup --format json-pretty

hints = 1
{
...
}
```
the 'hints = 1' cause yaml.load() to fail so I tried to handle it but it seems like this is failing in your case.
So, let's handle it specifically and in case we will see more issues with running ceph specific command we will not execute those from exec_oc_command() and create a dedicated func for that.
For now please try to replace the commented code with:
```python
        try:
            if out.startswith('hints = '):
                    out = out[out.index('{'):]
        except ValueError:
            pass
```

Reply:
Nice! This solution works in my case